File: csharp_model_trainer/CarValuePrediction/Services/CarValueModelTrainer.cs
using CSharpModelTrainer.SharedKernel;
using Microsoft.ML;
using SharedCL.CarValuePrediction.Mappings;
using SharedCL.CarValuePrediction.Models;
using System;
using System.Collections.Generic;
using System.Dynamic;
using System.Text;

namespace CSharpModelTrainer.CarValuePrediction.Services
{
    public class CarValueModelTrainer : IModelTrainer
    {
        public void TrainModel()
        {
            Console.WriteLine("=== Car Price Prediction Model Trainer === ");
            Console.WriteLine("=== Language: C# ===");
            Console.WriteLine();

            MLContext mlContext = new();
            var repoRoot = Path.GetFullPath(Path.Combine(AppDomain.CurrentDomain.BaseDirectory, "..", "..", "..", ".."));

            // load data
            Console.WriteLine("Loading data...");
            var dataPath = Path.Combine(repoRoot, "data", "car-prediction", "train.csv");
            IDataView data = mlContext.Data.LoadFromTextFile<CarInfoInput>(dataPath, hasHeader: true, separatorChar: ',');
            var allRows = mlContext.Data.CreateEnumerable<CarInfoInput>(data, reuseRowObject: false).ToList();
            Console.WriteLine("Sample data:");
            foreach (var item in allRows.Take(5))
            {
                Console.WriteLine(item);
            }
            Console.WriteLine($"Number of rows: {allRows.Count}");

            // filter data
            Console.WriteLine("Filtering data...");
            Console.WriteLine($"Rows before filtering: {allRows.Count}");
            IDataView filteredData = mlContext.Data.FilterRowsByColumn(data, "Price", lowerBound: 100, upperBound: 200000);
            var filteredCount = mlContext.Data.CreateEnumerable<CarInfoInput>(filteredData, reuseRowObject: false).Count();
            Console.WriteLine($"Rows after filtering: {filteredCount}");


            // split data
            var split = mlContext.Data.TrainTestSplit(filteredData, testFraction: 0.2, seed: 1);
            var trainSet = mlContext.Data
                .CreateEnumerable<CarInfoInput>(split.TrainSet, reuseRowObject: false);

            var testSet = mlContext.Data
                .CreateEnumerable<CarInfoInput>(split.TestSet, reuseRowObject: false);

            var trainSetDV = mlContext.Data.LoadFromEnumerable<CarInfoInput>(trainSet);
            var testSetDV = mlContext.Data.LoadFromEnumerable<CarInfoInput>(testSet);

            // data cleanup
            var mileageCleaner = new MileageCleanerMapping();
            var pipeline = mlContext.Transforms.CustomMapping(mileageCleaner.GetMapping(), contractName: "MileageCleaner")
                .Append(mlContext.Transforms.DropColumns("MileageKm"));


            // one hot encoding
            var multiColumnKeyPipeline =
                    mlContext.Transforms.Categorical.OneHotEncoding(
                        [
                            new InputOutputColumnPair("Manufacturer"),
                            new InputOutputColumnPair("Model"),
                            new InputOutputColumnPair("Category"),
                            new InputOutputColumnPair("LeatherInterior"),
                            new InputOutputColumnPair("FuelType"),
                            new InputOutputColumnPair("EngineVolume"),
                            new InputOutputColumnPair("GearBoxType"),
                            new InputOutputColumnPair("DriveWheels"),
                            new InputOutputColumnPair("Doors"),
                            new InputOutputColumnPair("Wheel"),
                            new InputOutputColumnPair("Color")
                        ]);

            var fullPipeline = pipeline
                .Append(multiColumnKeyPipeline)
                .Append(mlContext.Transforms.Concatenate("Features", [
                    "Manufacturer",
                    "Model",
                    "ProdYear",
                    "Category",
                    "LeatherInterior",
                    "FuelType",
                    "EngineVolume",
                    "Mileage",
                    "Cylinders",
                    "GearBoxType",
                    "DriveWheels",
                    "Doors",
                    "Wheel",
                    "Color",
                    "Airbags"
                ]))
                .Append(mlContext.Regression.Trainers.FastForest(
                    labelColumnName: "Price",
                    featureColumnName: "Features",
                    numberOfTrees: 200,
                    numberOfLeaves: 4096,
                    minimumExampleCountPerLeaf: 1
                ));

            Console.WriteLine("Training model...");
            var model = fullPipeline.Fit(trainSetDV);

            var modelDir = Path.Combine(repoRoot, "models", "car-prediction", "csharp");
            if (!Directory.Exists(modelDir))
            {
                Directory.CreateDirectory(modelDir);
            }

            var modelPath = Path.Combine(modelDir, "csharp_rf_carprice.zip");
            mlContext.Model.Save(model, trainSetDV.Schema, modelPath);

            // === Model Evaluation ===
            var trainPredictions = model.Transform(trainSetDV);
            var testPredictions = model.Transform(testSetDV);

            var trainMetrics = mlContext.Regression.Evaluate(trainPredictions, labelColumnName: "Price");
            var testMetrics = mlContext.Regression.Evaluate(testPredictions, labelColumnName: "Price");

            Console.WriteLine("┌─── TRAINING SET METRICS ───┐");
            Console.WriteLine($"  R²:   {trainMetrics.RSquared:F4}");
            Console.WriteLine($"  RMSE: {trainMetrics.RootMeanSquaredError:F2}");
            Console.WriteLine($"  MAE:  {trainMetrics.MeanAbsoluteError:F2}");
            Console.WriteLine("└─────────────────────────────┘");
            Console.WriteLine();
            Console.WriteLine("┌─── TEST SET METRICS ───┐");
            Console.WriteLine($"  R²:   {testMetrics.RSquared:F4}");
            Console.WriteLine($"  RMSE: {testMetrics.RootMeanSquaredError:F2}");
            Console.WriteLine($"  MAE:  {testMetrics.MeanAbsoluteError:F2}");
            Console.WriteLine("└─────────────────────────┘");
        }
    }
}

File: python-model-trainer/CarValuePrediction/car_value_model_trainer.py


import pandas as pd
import numpy as np
from sklearn.ensemble import RandomForestRegressor
from sklearn.model_selection import train_test_split
from sklearn.metrics import r2_score, root_mean_squared_error, mean_absolute_error
from pathlib import Path

repo_root = Path("..")

print("=== Car Price Prediction Model Trainer ===")
print("=== Language: Python ===")

# load data
print("Loading data...")
dataPath = repo_root.joinpath('data/car-prediction/train.csv')
data_df = pd.read_csv(dataPath)
print("Sample data:")
print(data_df.head())
print(f"Number of rows: {len(data_df)}")


# filter data
print("Filtering data...")
print(f"Rows before filtering: {len(data_df)}")
data_df = data_df[(data_df['Price'] >= 100) & (data_df['Price'] <= 200000)]
print(f"Rows after filtering: {len(data_df)}")


# data cleanup
data_df['Mileage'] = data_df['Mileage'].str.replace(' km', '').astype('float32')

# one hot encoding
columns = [col for col in data_df.columns if col not in ['ID', 'Levy']]
data_encoded_df = pd.get_dummies(data_df[columns])

# shuffle
# data_encoded_df = data_encoded_df.sample(frac=1).reset_index(drop=True)

feature_cols = [col for col in data_encoded_df.columns if col not in ['Price']]
X = data_encoded_df[feature_cols].to_numpy(dtype=np.float32)
y = data_encoded_df['Price'].to_numpy(dtype=np.float32)

# split data
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=1)


# train model
print("Training model...")
regr = RandomForestRegressor(
        n_estimators=200, 
        max_depth=25,
        random_state=1)

regr.fit(X_train, y_train)


# Saving Model
from skl2onnx import to_onnx

model_dir = repo_root / "models" / "car-prediction" / "python"
model_dir.mkdir(parents=True, exist_ok=True)

onx = to_onnx(regr, X[:1])
model_path = model_dir / "python_rf_carprice.onnx"
with open(model_path, "wb") as f:
    f.write(onx.SerializeToString())


# Model Evaluation
train_pred = regr.predict(X_train)
test_pred = regr.predict(X_test)

print("┌─── TRAINING SET METRICS ───┐")
print(f"  R2:   {r2_score(y_train, train_pred):.4f}")
print(f"  RMSE: {root_mean_squared_error(y_train, train_pred):.2f}")
print(f"  MAE:  {mean_absolute_error(y_train, train_pred):.2f}")
print("└─────────────────────────────┘")
print()
print("┌─── TEST SET METRICS ───┐")
print(f"  R2:   {r2_score(y_test, test_pred):.4f}")
print(f"  RMSE: {root_mean_squared_error(y_test, test_pred):.2f}")
print(f"  MAE:  {mean_absolute_error(y_test, test_pred):.2f}")
print("└─────────────────────────┘")

